README.md
# HoneyGain.js

[![axios](https://img.shields.io/github/package-json/dependency-version/LockBlock-dev/honeygain.js/axios)](https://www.npmjs.com/package/axios)

[![GitHub stars](https://img.shields.io/github/stars/LockBlock-dev/honeygain.js.svg)](https://github.com/LockBlock-dev/honeygain.js/stargazers) ![npm](https://img.shields.io/npm/dm/honeygain.js)

honeygain.js is a Node.js module that allows you to easily interact with the EarnApp API.

-   Promise-based
-   Performant
-   100% coverage of the HoneyGain API

## Installation

-   Install [NodeJS](https://nodejs.org).

### With GitHub:

-   Download or clone the project.
-   Go to the `honeygain.js` folder and run `npm install`.
-   Require [`client.js`](/src/client.js).

### With npm:

-   Run `npm install honeygain.js`.
-   Require the library.

## Documentation

-   [API documentation](/API.md)
-   [Changelog](/CHANGELOG.md)
-   [How to login with Authorization token](/Authorization.md)

## Example usage

The library can be used in both CommonJS and ES Modules

### Using the library

The library is async, be sure to use [async functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function#syntax) or [`.then()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then#syntax)

```js
const { Client } = require("honeygain.js");
//OR
import { Client } from "honeygain.js";

const client = new Client();

client.login("eyJ0eXAi...xaL5x4");

client.stats().then((data) => {
    console.log(data);
});
//OR
const getStats = async () => {
    const data = await client.stats();
    console.log(data);
};

getStats();
```

## Credits

[HoneyGain](https://honeygain.com)

## Copyright

See the [license](/LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.3.0] - 2022-08-05

### Added

-   added coupon function

### Changed

-   fixed notificationsAction missing parameter

## [1.2.0] - 2022-08-03

### Added

-   added contestWinnings function
-   added notificationsAction function

## [1.1.0] - 2022-08-02

### Added

-   added devices activity function

### Changed

-   dependencies update

## [1.0.1] - 2022-05-23

### Changed

-   removed require package.json at each request

## [1.0.0] - 2022-02-16

### Added

-   First version


src/errors.js
class BaseError extends Error {
    /**
     * Represents the base error.
     * @private
     * @param {string} type error type
     * @param {string} message error message
     */
    constructor(type, message) {
        super(message);

        /**
         * error type
         * @type {string}
         */
        this.type = type;
    }

    toJSON() {
        return {
            type: this.type,
            message: this.message,
        };
    }
}

class FatalError extends BaseError {
    /**
     * Represents a fatal error from the Client.
     * @extends BaseError
     * @param {string | error} error error object or message
     */
    constructor(error) {
        let errObject = typeof error == "string" ? null : error,
            message = errObject ? errObject.message : error;

        super("FatalError", message);

        if (errObject) this.stack = errObject.stack;
    }
}

class APIError extends BaseError {
    /**
     * Represents an error from the API.
     * @extends BaseError
     * @param {string | error} error error message
     * @param {object} response error response
     * @param {string} status status type of the request
     * @param {string} method method used for the request
     * @param {string} url url of the request to the endpoint
     */
    constructor(error, response, status, method, url) {
        let errObject = typeof error == "string" ? null : error,
            message = errObject ? errObject.message : error;

        super("APIError", message);

        /**
         * status type of the request
         * @type {string}
         */
        this.status = status;

        /**
         * method used for the request
         * @type {string}
         */
        this.method = method;

        /**
         * url of the request to the endpoint
         * @type {string}
         */
        this.url = url;

        if (response.data) {
            typeof response.data === "object"
                ? Object.assign(this, response.data)
                : (this.error = response.data);
        }
    }
}

class ParseError extends BaseError {
    /**
     * Represents a parsing error.
     * @extends BaseError
     * @param {String} message error message
     * @param {String} status status type of the request
     * @param {String} method method used for the request
     * @param {String} url url of the request to the endpoint
     */
    constructor(message, status, method, url) {
        super("ParseError", message);

        /**
         * status type of the request
         * @type {string}
         */
        this.status = status;

        /**
         * method used for the request
         * @type {string}
         */
        this.method = method;

        /**
         * url of the request to the endpoint
         * @type {string}
         */
        this.url = url;
    }
}

module.exports = { FatalError, APIError, ParseError };


src/client.js
const axios = require("axios").default;
const errors = require("./errors");
const pkg = require("../package.json");

class Client {
    /**
     * The HoneyGain client
     * @example const client = new Client();
     */
    constructor() {
        /**
         * The base API URL
         * @type {string}
         */
        this.API_URL = "https://dashboard.honeygain.com/api";
    }

    /**
     * Authentication token
     * @type {string}
     * @private
     */
    #authorization = "";

    /**
     * Make request against the API
     * @private
     * @param {string} method HTTP method
     * @param  {string} path API endpoint
     * @param  {Object} [reqOptions] request options
     * @private
     * @returns {Promise<Object>} promise
     */
    #request(method, path, reqOptions = {}) {
        let options = {
            method,
            url: `${this.API_URL}/${path}`,
            headers: {
                "User-Agent": `honeygain.js ${pkg.version} (https://github.com/LockBlock-dev/honeygain.js)`,
                "Content-Type": "application/json",
                "Accept-Encoding": "UTF8",
                Authorization: this.#authorization,
            },
            ...reqOptions,
        };

        return axios(options)
            .then((response) => {
                if (typeof response.data === "object") {
                    return response.data;
                } else {
                    try {
                        return { status: response.data.toLowerCase() };
                    } catch (err) {
                        throw new errors.ParseError(
                            response.data,
                            response.status,
                            options.method,
                            options.url
                        );
                    }
                }
            })
            .catch((error) => {
                throw error.type === "ParseError"
                    ? error
                    : new errors.APIError(
                          error,
                          error.response,
                          error.response.status,
                          options.method,
                          options.url
                      );
            });
    }

    /**
     * Log into HoneyGain.
     * @param {string} authorization Authorization token
     * @example client.login("eyJ0eXAi...xaL5x4");
     * @returns {Promise<Object>}
     */
    login(authorization) {
        this.#authorization = authorization;

        if (!this.#authorization.includes("Bearer"))
            this.#authorization = `Bearer ${this.#authorization}`;

        return { status: "ok" };
    }

    /**
     * Get your user data.
     * @example client.me();
     * @returns {Promise<Object>}
     */
    me() {
        return this.#request("GET", "v1/users/me");
    }

    /**
     * Get your tos status.
     * @example client.tos();
     * @returns {Promise<Object>}
     */
    tos() {
        return this.#request("GET", "v1/users/tos");
    }

    /**
     * Get user devices.
     * @example client.devices();
     * @returns {Promise<Object>}
     */
    devices() {
        return this.#request("GET", "v2/devices");
    }

    /**
     * Get user payouts.
     * @example client.payouts();
     * @returns {Promise<Object>}
     */
    payouts() {
        return this.#request("GET", "v1/payouts");
    }

    /**
     * Get user transactions.
     * @example client.transactions();
     * @returns {Promise<Object>}
     */
    transactions() {
        //to do: add date_from date_to
        return this.#request("GET", "v1/transactions_history");
    }

    /**
     * Get user stats.
     * @example client.stats();
     * @returns {Promise<Object>}
     */
    stats() {
        //to do: add date_from date_to (also for jmptstats)
        return this.#request("GET", "v1/earnings/stats");
    }

    /**
     * Get user referrals.
     * @example client.referrals();
     * @returns {Promise<Object>}
     */
    referrals() {
        return this.#request("GET", "v1/referrals");
    }

    /**
     * Get user balances.
     * @example client.balances();
     * @returns {Promise<Object>}
     */
    balances() {
        return this.#request("GET", "v1/users/balances");
    }

    /**
     * Get user notifications.
     * @example client.notifications();
     * @returns {Promise<Object>}
     */
    async notifications() {
        const { data } = await this.me();
        return this.#request("GET", `v1/notifications?user_id=${data.id}`);
    }

    /**
     * Get referrals earnings.
     * @example client.referralsEarnings();
     * @returns {Promise<Object>}
     */
    referralsEarnings() {
        return this.#request("GET", "v1/referrals/earnings");
    }

    /**
     * Get user Jump Task earnings.
     * @example client.jumpTaskEarnings();
     * @returns {Promise<Object>}
     */
    jumpTaskEarnings() {
        return this.#request("GET", "v1/earnings/jt");
    }

    /**
     * Get user Jump Task stats.
     * @example client.jumpTaskStats();
     * @returns {Promise<Object>}
     */
    jumpTaskStats() {
        return this.#request("GET", "v1/jt-earnings/stats");
    }

    /**
     * Get user jumpTask earnings for the day.
     * @example client.today();
     * @returns {Promise<Object>}
     */
    jumpTaskToday() {
        return this.#request("GET", "v1/jt-earnings/today");
    }

    /**
     * Get user wallet stats.
     * @example client.walletStats();
     * @returns {Promise<Object>}
     */
    walletStats() {
        return this.#request("GET", "v1/earnings/wallet-stats");
    }

    /**
     * Get user earnings for the day.
     * @example client.today();
     * @returns {Promise<Object>}
     */
    today() {
        return this.#request("GET", "v1/earnings/today");
    }

    /**
     * Get user devices activity.
     * @param {Object} [params = {}] - maximum time window of 30 days
     * @param {string} params.date_from - start date
     * @param {string} params.date_to - end date (maximum tomorrow)
     * @example client.devicesActivity({ date_from: "2022-08-01", date_to: "2022-08-05" });
     * @returns {Promise<Object>}
     */
    devicesActivity(params = {}) {
        return this.#request("GET", `v2/devices/activity?${new URLSearchParams(params)}`);
    }

    /**
     * Change user password.
     * @param {string} currentPassword your current HoneyGain password
     * @param {string} newPassword your new HoneyGain password
     * @example client.changePassword("myOldPass", "myNewPass");
     * @returns {Promise<Object>}
     */
    changePassword(currentPassword, newPassword) {
        return this.#request("PUT", "v1/users/passwords", {
            data: { current_password: currentPassword, new_password: newPassword },
        });
    }

    /**
     * Delete your account.
     * You will need to click on the link in the mail sent to your honeygain email!
     * @example client.deleteAccount();
     * @returns {Promise<Object>}
     */
    deleteAccount() {
        return this.#request("POST", "v1/users/deletion_requests");
    }

    /**
     * Execute a notifications action.
     * @param {string} notificationHash - notification hash
     * @param {Object} [body = {}] - action
     * @example client.notificationsAction({ campaign_id: "someID", action: "someAction", user_id: "userID" });
     * @returns {Promise<Object>}
     */
    async notificationsAction(notificationHash, body = {}) {
        return this.#request("POST", `v1/notifications/${notificationHash}/actions`, {
            data: body,
        });
    }

    /**
     * Get your contest winnings (once a day).
     * @example client.contestWinnings();
     * @returns {Promise<Object>}
     */
    contestWinnings() {
        return this.#request("POST", "v1/contest_winnings");
    }

    /**
     * Get information about a coupon/referral code.
     * @param {string} code - coupon code
     * @example client.coupon("freemoney");
     * @return {Promise<Object>}
     */
    coupon(code) {
        return this.#request("GET", `v1/coupons/${code}`);
    }

    // /**
    //  * Redeem your balance.
    //  * @param {string} email paypal / amazon email
    //  * @param {string} method methods:
    //  * to see methods do client.paymentMethods()
    //  * @example client.redeem("mail@domain.com", "paypal.com");
    //  * @example client.redeem("mail@domain.com", "amazon.com (US/Global)");
    //  * @returns {Promise<Object>}
    //  */
    // redeem(email, method) {
    //     return this.#request("POST", "redeem", { data: { to: email, payment_method: method } });
    // }
}

module.exports = { Client };
